### ISSUE TITLE
sql,kv: investigate returning EncDatum format from KV on scans

### ISSUE THREAD

**Issue by RaduBerinde** [cockroachdb/cockroach#13193]
Since there is measurable overhead from serializing/deserializing keys/values we would likely get a benefit from a more efficient data encoding between the KV and SQL layers.

The idea (suggested by @petermattis) is to break apart the encoded keys/values (using `encoding.PeekLength` and such) and return rows of `EncDatum`s (or something that directly maps to `EncDatum`s) directly from KV. KV would understand how to parse values (column families etc); there would be a special mode or variant of ScanRequest which has the key and value types passed down from SQL.

This would also allow us to avoid sending out data for columns the scan doesn't need.

**Comment by petermattis** [comment]
I believe the major inefficiency in the existing `ScanResponse` is our use of protos:

```
message Value {
  optional bytes raw_bytes = 1;
  optional util.hlc.Timestamp timestamp = 2;
}

message KeyValue {
  optional bytes key;
  optional Value value;
}

message ScanResponse {
  repeated KeyValue rows;
}
```

When we unmarshal a `ScanResponse`, each `KeyValue` requires 2 allocations: one for the key and one for the value. There is also a small bit of overhead in the protobuf encoding. My thinking is that a flag in the `ScanRequest` could indicate an alternate return format that is composed of SQL rows, encoded compactly:

```
message ScanResponse {
  ...
  optional group sql_rows {
    repeated int32 row_starts;
    required bytes data;
  }
}
```

`sql_rows.row_starts` would be the index into `sql_rows.data` of the start of each row. Each row would be a series of `EncDatums`. Lots more details to figure out here such as how to specify in the `ScanRequest` how to combine multiple key/values into a single row and which columns to return, but none of that seems insurmountable.

**Comment by RaduBerinde** [comment]
I think this would improve performance significantly for most queries, especially when we have "wide" tables. I will work on a proposal for the API.

**Comment by tschottdorf** [comment]
You could also make the change Peter suggested and then see how it improves the actual performance.

**Comment by RaduBerinde** [comment]
What change is that exactly? There are a lot of details missing from the rough proposal above. I want to come up with a complete proposal before I go ahead with all the implementation (which is anything but trivial).

**Comment by tschottdorf** [comment]
Sorry, didn't mean to say don't make a proposal. Meant to say don't make a full-fledged proposal before you've managed to get some sanity checks that this is useful. That's likely obvious to you anyway, so sorry about the noise.

**Comment by RaduBerinde** [comment]
Fair enough. I am also looking from the perspective of DistSQL where the scan is (mostly) local so there is less concern about masrhal/unmarshal overhead.  It's less clear that there would be a big win in that case.

**Comment by jordanlewis** [comment]
I fixed this in #26553.
